pkg/agent/resourcemanager/fetcher/kubelet/topology/topology_adapter.go

```diff
@@ -449,6 +498,73 @@ func (p *topologyAdapterImpl) getZoneAllocations(podList []*v1.Pod, podResources
 	return zoneAllocationsMap, nil
 }
 
+func (p *topologyAdapterImpl) getDeviceZoneAllocations(podList []*v1.Pod, podResourcesList []*podresv1.PodResources) (map[util.ZoneNode]util.ZoneAllocations, error) {
+	var (
+		err     error
+		errList []error
+	)
+
+	podMap := native.GetPodNamespaceNameKeyMap(podList)
+	zoneAllocationsMap := make(map[util.ZoneNode]util.ZoneAllocations)
+	for _, podResources := range podResourcesList {
+		if podResources == nil {
+			continue
+		}
+
+		podKey := native.GenerateNamespaceNameKey(podResources.Namespace, podResources.Name)
+		pod, ok := podMap[podKey]
+		if !ok {
+			errList = append(errList, fmt.Errorf("pod %s not found in metaserver", podKey))
+			continue
+		}
+
+		if native.PodIsTerminated(pod) {
+			continue
+		}
+
+		// the pod resource filter will filter out unwanted pods
+		if p.podResourcesFilter != nil {
+			podResources, err = p.podResourcesFilter(pod, podResources)
+			if err != nil {
+				errList = append(errList, err)
+				continue
+			}
+
+			// if podResources is nil, it means that the pod is filtered out
+			if podResources == nil {
+				continue
+			}
+		}
+
+		for _, c := range podResources.Containers {
+			for _, device := range c.Devices {
+				for targetResourceName, targetZoneName := range p.resourceNameToZoneNameMap {
+					if device.ResourceName == targetResourceName {
+						for _, deviceId := range device.DeviceIds {
+							deviceNode := util.GenerateDeviceZoneNode(deviceId, targetZoneName)
+							if _, ok := zoneAllocationsMap[deviceNode]; !ok {
+								zoneAllocationsMap[deviceNode] = []*nodev1alpha1.Allocation{
```

Comment:
all this logic can implement in addContainerDevices, because the zoneNode should also need have allocatable and capacity when allocation has request

Reply:
done